common/djangoapps/third_party_auth/pipeline.py

```diff
@@ -445,13 +427,46 @@ def parse_query_params(strategy, response, *args, **kwargs):
     """Reads whitelisted query params, transforms them into pipeline args."""
     auth_entry = strategy.session.get(AUTH_ENTRY_KEY)
     if not (auth_entry and auth_entry in _AUTH_ENTRY_CHOICES):
-        raise AuthEntryError(strategy.backend, 'auth_entry missing or invalid')
+        raise AuthEntryError(strategy.request.backend, 'auth_entry missing or invalid')
 
     return {'auth_entry': auth_entry}
 
 
+def set_pipeline_timeout(strategy, user, *args, **kwargs):
+    """
+    Set a short session timeout while the pipeline runs, to improve security.
+
+    Consider the following attack:
+    1. Attacker on a public computer visits edX and initiates the third-party login flow
+    2. Attacker logs into their own third-party account
+    3. Attacker closes the window and does not complete the login flow
+    4. Victim on the same computer logs into edX with username/password
+    5. edX links attacker's third-party account with victim's edX account
+    6. Attacker logs into victim's edX account using attacker's own third-party account
+
+    We have two features of the pipeline designed to prevent this attack:
+    * This method shortens the Django session timeout during the pipeline. This should mean that
+      if there is a reasonable delay between steps 3 and 4, the session and pipeline will be
+      reset, and the attack foiled.
+      Configure the timeout with the SOCIAL_AUTH_PIPELINE_TIMEOUT setting (Default: 600 seconds)
+    * On step 4, the login page displays an obvious message to the user, saying "You've
+      successfully signed into (Google), but your (Google) account isn't linked with an edX
+      account. To link your accounts, login now using your edX password.".
+    """
+    if strategy.request and not user:  # If user is set, we're currently logged in so it doesn't matter.
+        strategy.request.session.set_expiry(strategy.setting('PIPELINE_TIMEOUT', 600))
+        strategy.request.session['third_party_pipeline_timeout_active'] = True
+
+
+def reset_pipeline_timeout(strategy, *args, **kwargs):
+    """ Revert the short session timeout now that the pipeline is completed. """
+    if strategy.request and strategy.request.session.pop('third_party_pipeline_timeout_active', False):
+        strategy.request.session.set_expiry(None)  # None makes the session revert to the global session expiry policy.
```

Comment:
Thanks @nasthagiri, good catch. It turns out that we don't need this `reset_pipeline_timeout` step at all - because we only set a shortened timeout if no user is logged in and no edx user is linked to the third party account. In that case, the user must necessarily either do a normal login or register a new account. Either of those actions will set a new expiry time, as you've just pointed out.

I've pushed a fix and tested it.